0000-lossy-conversions.md

```diff
@@ -0,0 +1,310 @@
+- Feature Name: `lossy_conversions`
+- Start Date: 2023-04-14
+- RFC PR: [rust-lang/rfcs#3415](https://github.com/rust-lang/rfcs/pull/3415)
+- Rust Issue: TBD <!-- [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000) -->
+
+# Summary
+
+[summary]: #summary
+
+Add traits for lossy numeric conversions as an alternative to the `as` operator, and deprecate `as` for lossy numeric casts in a future edition, so
+
+```rust
+let n = f64::PI as usize;
+```
+
+becomes
+
+```rust
+let n: usize = f64::PI.approx();
+// or
+let n = f64::PI.approx::<usize>();
+```
+
+# Motivation
+
+[motivation]: #motivation
+
+The `as` operator is a footgun when used to convert between number types. For example, converting an `i64` to an `i32` may silently truncate digits. Other conversions may wrap around, saturate, or lose numerical precision.
+
+The problem is that this is not obvious when reading the code; the `as` operator looks innocuous, so the risk of getting a wrong result is easily overlooked. This goes against Rust's design philosophy of highlighting potential problems with explicit syntax. This is similar to the `unsafe` keyword, which makes unsafe Rust more verbose, but also highlights code that could cause Undefined Behaviour. `as` can not introduce UB, but it can be a logic error. Rust also tries to prevent logic errors, e.g. by requiring that a `match` covers all possible values, and errors aren't silently ignored.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Conversions between number types can sometimes be lossy. For example, when converting a `f64` to `f32`, the number gets less precise, and very large numbers become _Infinity_. Rust offers the following methods for converting between numbers:
+
+- `into()`: This covers _lossless_ conversions, where the output is the exact same number as the input. For example, converting a `u8` to a `u32` is lossless, because every value of `u8` can be represented by a `u32`. You may also use the `as` operator for lossness numeric conversions.
```

Comment:
`as` cannot be deprecated completely, and I don't think it makes sense anyway. Besides numeric casts, `as` is used for many kinds of type coercion, including unsizing coercions, deref coercion, enum to discriminant coercions, raw pointer casts, etc. Deprecating `as` for _lossy_ conversions is about removing a footgun, but I do not see a compelling motivation for deprecating `as` altogether. It would cause a lot of churn for no apparent benefit.

Reply:
By "completely" I was referring to lossless, not type coercions, but I see you have covered that in the bototm section as well.